swarm/src/connection.rs

```diff
@@ -0,0 +1,204 @@
+            .finish()
+    }
+}
+
+impl<TMuxer, THandler> Unpin for Connection<TMuxer, THandler>
+where
+    TMuxer: StreamMuxer,
+    THandler: ConnectionHandler<Substream = Substream<TMuxer>>,
+{
+}
+
+impl<TMuxer, THandler> Connection<TMuxer, THandler>
+where
+    TMuxer: StreamMuxer,
+    THandler: ConnectionHandler<Substream = Substream<TMuxer>>,
+{
+    /// Builds a new `Connection` from the given substream multiplexer
+    /// and connection handler.
+    pub fn new(muxer: TMuxer, handler: THandler) -> Self {
+        Connection {
+            muxing: Muxing::new(muxer),
+            handler,
+        }
+    }
+
+    /// Notifies the connection handler of an event.
+    pub fn inject_event(&mut self, event: THandler::InEvent) {
+        self.handler.inject_event(event);
+    }
+
+    /// Begins an orderly shutdown of the connection, returning the connection
+    /// handler and a `Future` that resolves when connection shutdown is complete.
+    pub fn close(self) -> (THandler, Close<TMuxer>) {
+        (self.handler, self.muxing.close().0)
+    }
+
+    /// Polls the connection for events produced by the associated handler
+    /// as a result of I/O activity on the substream multiplexer.
+    pub fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Event<THandler::OutEvent>, ConnectionError<THandler::Error>>> {
+        loop {
+            let mut io_pending = false;
+
+            // Perform I/O on the connection through the muxer, informing the handler
+            // of new substreams.
+            match self.muxing.poll(cx) {
+                Poll::Pending => io_pending = true,
+                Poll::Ready(Ok(SubstreamEvent::InboundSubstream { substream })) => self
+                    .handler
+                    .inject_substream(substream, SubstreamEndpoint::Listener),
+                Poll::Ready(Ok(SubstreamEvent::OutboundSubstream {
+                    user_data,
+                    substream,
+                })) => {
+                    let endpoint = SubstreamEndpoint::Dialer(user_data);
+                    self.handler.inject_substream(substream, endpoint)
+                }
+                Poll::Ready(Ok(SubstreamEvent::AddressChange(address))) => {
+                    self.handler.inject_address_change(&address);
+                    return Poll::Ready(Ok(Event::AddressChange(address)));
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(ConnectionError::IO(err))),
+            }
+
+            // Poll the handler for new events.
+            match self.handler.poll(cx) {
+                Poll::Pending => {
+                    if io_pending {
+                        return Poll::Pending; // Nothing to do
+                    }
+                }
+                Poll::Ready(Ok(ConnectionHandlerEvent::OutboundSubstreamRequest(user_data))) => {
+                    self.muxing.open_substream(user_data);
```

Comment:
>  this is cooperative scheduling after all

Yes, good point.

> poll is called deep inside some nested task

This is running on the connection specific task. In other words, there is only one level above it, namely the future polling the channel to the `Swarm` and `Connection::poll`.

(Note that this has not been changed in this pull request, merely copied from `core/src/connection.rs`. Not arguing to not change it, just suggesting to change it within another pull request.)

Reply:
Right, so in this case `wake_by_ref` is desirable. This is one unfortunate consequence of Rust’s `Future` design: reusing this future somewhere else may break this assumption, which will silently introduce performance or fairness issues.

Reply:
In order for `wake_by_ref` to work, the following guarantee needs to be made by an executor:

_If the waker of a task is woken, where the task itself is currently running, the task is still rescheduled to be polled again._

In other words, the optimization of ignoring wake signals for running tasks needs to be forbidden.

If I am not mistaken this is an unwritten rule, or do you know of a place where this behavior is documented @rkuhn? Maybe @tomaka knows this?

Reply:
Hmm, that wouldn’t be an optimisation, it would break a lot of code. [Including Tokio](https://github.com/tokio-rs/tokio/blob/master/tokio/src/coop.rs#L156-L182).

Reply:
opened https://github.com/rust-lang/rust/issues/93961

Reply:
Somewhat related and an interesting read: https://tokio.rs/blog/2020-04-preemption